pkg/api/soc.go

```diff
@@ -42,133 +42,122 @@ func (s *Service) socUploadHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	headers := struct {
+		BatchID []byte `map:"Swarm-Postage-Batch-Id" validate:"required"`
+		Pin     bool   `map:"Swarm-Pin"`
+	}{}
+	if response := s.mapStructure(r.Header, &headers); response != nil {
+		response("invalid header params", logger, w)
+		return
+	}
+
+	tag, err := s.storer.NewSession()
+	if err != nil {
+		logger.Debug("get or create tag failed", "error", err)
+		logger.Error(nil, "get or create tag failed")
+		switch {
+		case errors.Is(err, storage.ErrNotFound):
+			jsonhttp.NotFound(w, "tag not found")
+		default:
+			jsonhttp.InternalServerError(w, "cannot get or create tag")
+		}
+		return
+	}
+
+	putter, err := s.newStamperPutter(r.Context(), putterOptions{
+		BatchID:  headers.BatchID,
+		TagID:    tag.TagID,
+		Pin:      headers.Pin,
+		Deferred: true,
+	})
+	if err != nil {
+		logger.Debug("get putter failed", "error", err)
+		logger.Error(nil, "get putter failed")
+		switch {
+		case errors.Is(err, errBatchUnusable) || errors.Is(err, postage.ErrNotUsable):
+			jsonhttp.UnprocessableEntity(w, "batch not usable yet or does not exist")
+		case errors.Is(err, postage.ErrNotFound):
+			jsonhttp.NotFound(w, "batch with id not found")
+		case errors.Is(err, errInvalidPostageBatch):
+			jsonhttp.BadRequest(w, "invalid batch id")
+		default:
+			jsonhttp.BadRequest(w, nil)
+		}
+		return
+	}
+
+	ow := &cleanupOnErrWriter{ResponseWriter: w, onErr: putter.Cleanup}
```

Comment:
The logger is not injected. The same applies wherever `cleanupOnErrWriter` is used.